### hw  

### Проверка статистических гипотез. P-значения. Доверительные интервалы

__Задача 1__

Известно, что генеральная совокупность распределена нормально с известным средним квадратическим отклонением 16. Найти ширину доверительного интервала для оценки математического ожидания `mu` с надёжностью 0.95, если выборочное среднее равно 80, а объём выборки равен 256.
<hr>

<center>Решение: <br>

Доверительные интервалы используют ту же математическую базу, что и статистические тесты.

У нас дана выборка $X$ из __нормально распределённой__ случайной величины с __известной дисперсией__ $\sigma ^ 2$, и требуется построить доверительный интервал для математического ожидания $\mu$ с доверительной вероятностью $p$. Мы знаем, что в этом случае статистика
$$Z = \dfrac{\overline{X} - \mu}{\sigma / \sqrt{n}}$$
имеет стандартное нормальное распределение. 

Обозначим $\alpha = 1 - p$. Можно убедиться в том, что
$$P \left( t_{\alpha / 2} \leq Z \leq t_{1 - \alpha / 2} \right) = p,$$
где $t_{x}$ — квантиль порядка $x$ для стандартного нормального распределения. Подставляя сюда $Z$, получаем
$$P \left( t_{\alpha / 2} \leq \dfrac{\overline{X} - \mu}{\sigma / \sqrt{n}} \leq t_{1 - \alpha / 2} \right) = p$$
$$P \left( t_{\alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \leq \overline{X} - \mu \leq t_{1 - \alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \right) = p$$


<center>
    <div style="width: 350px;border:1px solid black;">
        $P \left( \overline{X} + t_{\alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \leq \mu \leq \overline{X} + t_{1 - \alpha / 2} \cdot \dfrac{\sigma}{\sqrt{n}} \right) = p$
    </div>
</center>

    

$\sigma = 16$ - средне-квадратическое отклонение<br>
$n = 256$ - объем выборки<br>
$\overline{X} = 80$ - выборочное среднее<br>
$p = 0.95$ - надежность<br>

In [6]:
import numpy as np
import scipy.stats as sts

In [7]:
X_ = 80 
mu = 1
sigm = 16 
n = 256

In [8]:
# Найдем квантили t1 t2
p = 0.95
alpha = 1 - p

t1 = sts.t.ppf(alpha / 2, df=n - 1)
t2 = sts.t.ppf(1 - alpha / 2, df=n - 1)

t1, t2

(-1.9693105698491928, 1.9693105698491928)

In [11]:
# Доверительный интервал
(X_ + t1 * sigm / np.sqrt(n), X_ + t2 * sigm / np.sqrt(n))

(78.0306894301508, 81.9693105698492)

In [15]:
# Ширина доверительного интервала:
t = ( X_ + t2 * sigm / np.sqrt(n)) - (X_ + t1 * sigm / np.sqrt(n))
t

3.938621139698398

<hr>

__Задача 2__

Продавец утверждает, что средний вес пачки печенья составляет 200 г. Из партии извлечена выборка из 10 пачек. Вес каждой пачки составляет:
```
202, 203, 199, 197, 195, 201, 200, 204, 194, 190
```
Известно, что их веса распределены нормально. Верно ли утверждение продавца, если учитывать, что уровень значимости равен 1%? Найдите P-значение для данного теста и укажите его в качестве ответа.
<hr>

<center>Решение: <br>
    
    
x_array = [202, 203, 199, 197, 195, 201, 200, 204, 194, 190]<br>
$n = 10$ - объем выборки<br>
$\mu = 200$<br>
$\alpha = 0.01$ - уровень значимости<br>

In [79]:
x_array = np.array([202, 203, 199, 197, 195, 201, 200, 204, 194, 190])
X = 200
n = x_array.shape[0]
p = 0.01
alpha = 1 - p

In [83]:
sigma = np.std(x_array, ddof=1)
Xmat = np.mean(x_array)
sigma, Xmat

(4.453463071962462, 198.5)

In [84]:
Zn = (Xmat - X) / (sigma / np.sqrt(n))
Zn

-1.0651074037450896

In [85]:
# Z в области гипотезы H0, следовательно H0 верна на уровне значимости 1%
p_left = sts.t.cdf(Zn, df=n - 1)
p_right = 1 - sts.t.cdf(Zn, df=n - 1)

pvalue = 2 * min(p_left, p_right)

pvalue

0.3145666035247285

<hr>

__Задача 3__

Дана выборка диаметров подшипников из примера 1 с занятия 5:
```
samples = [0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
           1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
           0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02  ]
```
Предполагая, что диаметры подшипников распределены нормально, проверьте гипотезу о том, что дисперсия случайной величины равна `0.0625` при уровне значимости `alpha = 0.05`. Что для этого нужно знать:
1. Альтернативная гипотеза двухсторонняя.
2. Статистика для теста: `H = (n - 1) * sample_variance / variance`, где `n` - число элементов в выборке, `sample_variance` - несмещённая оценка дисперсии, `variance` - утверждаемая нулевой гипотезой дисперсия.
3. Эта статистика в предположении верности нулевой гипотезы имеет распределение хи-квадрат с параметром `df = n - 1`. Её квантили можно найти с помощью функции `scipy.stats.chi2.ppf`.

В качестве ответа запишите 1, если нулевая гипотеза верна, и 0 иначе.

In [75]:
import numpy as np
import scipy.stats as sts

In [57]:
samples = [0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
           1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
           0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02  ]

In [87]:
np.mean(samples)

1.1084541666666665

In [88]:
samples = np.array([0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
           1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
           0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02])
alpha = 0.05
n = samples.shape[0]
D = 0.0625

t1 = sts.chi2.ppf(alpha / 2, df=n - 1)
t2 = sts.chi2.ppf(1 - alpha / 2, df=n - 1)
t1, t2

(11.688551922452438, 38.0756272503558)

In [70]:
# Найдем квантили t1 t2
alpha = 0.05

t1 = sts.chi2.ppf(alpha / 2, df=n - 1)
t2 = sts.chi2.ppf(1 - alpha / 2, df=n - 1)

t1, t2

(11.688551922452438, 38.0756272503558)

Итак, критическая область:
$$\Omega_\alpha = (-\infty, 11.69) \cup (38.08, \infty)$$

In [91]:
def statistic_3(samples):
    return (n - 1)*(np.var(samples, ddof=1))/D

In [92]:
statistic_3(samples)

28.720541753333343

Значение статистики не попало в критическую область, значит, гипотеза остаётся в силе.